File: DFS/312.Burst-Balloons/312.Burst-Balloons-DP.py
class Solution:
    def maxCoins(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        N = len(nums)
        nums = [1]+nums+[1]        
        dp = [[0 for _ in range(N+2)] for _ in range(N+2)]        
        for i in range(1,N+1):
            dp[i][i]=nums[i-1]*nums[i]*nums[i+1]  
        
        for Len in range(2,N+1):
            for i in range(1,N-Len+2):
                left,right = i,i+Len-1
                for k in range(left,right+1):
                    dp[left][right] = max(dp[left][right],dp[left][k-1]+nums[left-1]*nums[k]*nums[right+1]+dp[k+1][right])
                        
        return dp[1][N]

File: DFS/312.Burst-Balloons/312.Burst-Balloons-DP.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) 
    {
        int N = nums.size();
        nums.insert(nums.begin(),1);
        nums.push_back(1);
        
        auto dp = vector<vector<int>>(N+2,vector<int>(N+2,0));
        
        for (int len=1; len<=N; len++)
            for (int i=1; i+len-1<=N; i++)
            {
                int j = i+len-1;
                for (int k=i; k<=j; k++)
                    dp[i][j] = max(dp[i][j], dp[i][k-1]+dp[k+1][j]+nums[i-1]*nums[k]*nums[j+1]);                
            }        
        
        return dp[1][N];
    }
};
